File: leetcode/46.rs
/*
Question:
Given an array nums of distinct integers, return all the possible permutations. 
You can return the answer in any order.
*/

// Approach:
// The backtrack functions works as follows
// 1) The base case checls if the current index exceeds the length of the permutation
//    And pushes the current permutation to the ans vector if it does
// 2) Otherwise we loop from idx(current index) to the end of array
//    We swap the ith element with idx and then backtrack with the incremented index
//    After that we undo the swap and proceed to next iteration of the loop

impl Solution {
    fn backtrack(nums: &mut Vec<i32>, idx: usize, ans: &mut Vec<Vec<i32>>) {
        if idx >= nums.len() {
            ans.push(nums.clone());
            return;
        }

        for i in idx..nums.len() {
            nums.swap(i, idx);
            Self::backtrack(nums, idx + 1, ans);
            nums.swap(idx, i);
        }
    }
    pub fn permute(mut nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut ans = vec![];
        Self::backtrack(&mut nums, 0, &mut ans);
        ans
    }
}

File: leetcode/46.java
/*
Queestion:
Given an array nums of distinct integers, return all the possible permutations. You can return the answer in any order.
*/

// Approach:
// 1) Make a 2d List of Integers to store the permutations
// 2) Make and use generate() to generate the permutations
//    generate() will take an additional parameter called fixed, which is the index of last fixed element in the current permutation
//    If all numbers are fixed, i.e. fixed == nums.length-1, then add the current order of nums to permutations
//    Otherwise we loop over all non-fixed indices and fix each one of them one by one 
//    We do this by swapping them with the fixed index and calling generate to generate combinations with that element fixed
// 3) After the call to generate(), permutations will be updated and we can return it

class Solution {
    void generate(int[] nums, List<List<Integer>> permutations, int fixed) {
        if(fixed == nums.length-1) {
            List<Integer> list = new ArrayList<>();
            for(int num: nums) list.add(num);
            permutations.add(list);
            return;
        }
        for(int i=fixed; i<nums.length; i++) {
            int tmp = nums[fixed];
            nums[fixed] = nums[i];
            nums[i] = tmp;
            this.generate(nums,permutations,fixed+1);
            nums[i] = nums[fixed];
            nums[fixed] = tmp;
        }
    }
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> permutations = new ArrayList<List<Integer>>();
        this.generate(nums,permutations,0);
        return permutations;
    }
}
